language-server/test/dotty/tools/languageserver/util/CodeTester.scala

```diff
@@ -30,6 +33,32 @@ class CodeTester(projects: List[Project]) {
 
   private val positions: PositionContext = getPositions(files)
 
+  /** Check that the last diagnostics that have been published so far by the server
+   *  for a given file match `expected`.
+   *
+   * @param marker   The marker defining the source file from which to query.
+   * @param expected The expected diagnostics to be found
+   */
+  def diagnostics(marker: CodeMarker,
+      expected: (CodeRange, String, DiagnosticSeverity, Option[ErrorMessageID])*): this.type = {
+    implicit val posCtx = positions
+
+    def toDiagnostic(range: CodeRange, message: String, severity: DiagnosticSeverity,
+        errorCode: Option[ErrorMessageID]): Diagnostic = {
+      val location = range.toLocation
+      new Diagnostic(
+        location.getRange, message, severity, /*source=*/"", errorCode.map(_.errorNumber.toString).getOrElse(null)
+      )
+    }
+
+    val expectedParams = marker.toPublishDiagnosticsParams(expected.toList.map(toDiagnostic))
+    val actualParams = testServer.client.diagnostics.get.reverse.find(_.getUri == marker.uri)
```

Comment:
Instead of reverse, I'd suggest sorting the diagnostics by their `Location`:
```scala
  private implicit val LocationOrdering: Ordering[Location] = Ordering.by(_.toString)
```

Reply:
the reverse is here to sort the list of PublishDiagnosticsParams (every time the server calls publishDiagnostics, something gets appended to that list), but we could sort the list of diagnostics by location indeed.

Reply:
On second thoughts, this is a bit annoying to do since it requires constructing a new instance of PublishDiagnosticsParams, so I'll leave it in the original order for now.